lib/ts/recipe/session/recipe.ts

```diff
@@ -158,21 +153,53 @@ export default class Session extends RecipeModule<unknown, unknown, unknown, any
             });
         }
 
+        const globalValidators: SessionClaimValidator[] = await getGlobalClaimValidators();
+
         // We validate all the global claims
         const invalidClaims = await this.validateClaims({ userContext });
 
-        // Check if any of those claim errors requests a redirection
-        const invalidClaimRedirectPath = popInvalidClaimRedirectPathFromContext(userContext);
-        if (invalidClaims.length > 0 && invalidClaimRedirectPath !== undefined) {
+        if (invalidClaims.length > 0) {
             if (redirectInfo !== undefined) {
                 // if we have to redirect and we have success context we wanted to use we save it in localstorage
-                // this way after the other page did solved the validation error it can contine
+                // this way after the other page did solved the validation error it can continue
                 // the sign in process by calling this function without passing the redirect info
                 const jsonContext = JSON.stringify(redirectInfo);
                 await setLocalStorage("supertokens-success-redirection-context", jsonContext);
             }
-            // then we do the redirection.
-            return SuperTokens.getInstanceOrThrow().redirectToUrl(invalidClaimRedirectPath, history);
+
+            // we try to find claim validator among failed validators with onFailure cb that returns string
+            const getRedirectionPath = async () => {
+                const callbacks = [];
+                for (const claim of invalidClaims) {
+                    const validator = globalValidators.find(({ id }) => id === claim.validatorId);
+                    if (validator?.onFailureRedirection) {
+                        callbacks.push(validator?.onFailureRedirection?.());
+                    }
+                }
+                return (await Promise.all(callbacks)).filter(Boolean)[0];
+            };
+
+            const redirectPath = await getRedirectionPath();
+
+            // if none has onFailure that returns string we fallback to default
```

Comment:
If onFailure is defined but returns undefined we do not need to redirect to the access denied page. 
In that case, we can continue with the onSuccess redirections (for the non-failing claims ofc).